File: evals/tools/measure_skills.py
#!/usr/bin/env python3
"""Report compression stats for the embedded agent skills.

Installs BOTH permutations with the real renderer into temporary directories and
measures the flat files that actually ship.

Going through `magus agent install` rather than re-implementing applyVariant here
is deliberate: this script cannot drift from the Go renderer, and it keeps
working unchanged when the body syntax changes.

Run from the repo root.

Usage: python3 evals/tools/measure_skills.py [--json]
"""

import argparse
import glob
import json
import os
import shlex
import subprocess
import sys
import tempfile

MAGUS = os.environ.get("MAGUS_BIN", "go run ./cmd/magus")


def install(dest, simple):
    """Render one permutation into dest with the real renderer."""
    cmd = shlex.split(MAGUS) + ["agent", "install", "--dir", dest, "--force", ".claude/skills"]
    if simple:
        cmd.append("--simple")
    proc = subprocess.run(cmd, capture_output=True, text=True)
    if proc.returncode != 0:
        raise SystemExit("install failed: {}\n{}".format(" ".join(cmd), proc.stderr))
    return {
        os.path.basename(os.path.dirname(path)): open(path, encoding="utf-8").read()
        for path in sorted(glob.glob(os.path.join(dest, ".claude", "skills", "*", "SKILL.md")))
    }


def compose(text):
    """Byte counts by line category, for the prose vs hard-content split."""
    counts = {"prose": 0, "table": 0, "code": 0, "heading": 0, "blank": 0}
    in_fence = False
    for line in text.split("\n"):
        n = len(line) + 1
        stripped = line.strip()
        if stripped.startswith("```"):
            in_fence = not in_fence
            counts["code"] += n
        elif in_fence:
            counts["code"] += n
        elif stripped.startswith("|"):
            counts["table"] += n
        elif stripped.startswith("#"):
            counts["heading"] += n
        elif not stripped:
            counts["blank"] += n
        else:
            counts["prose"] += n
    return counts


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--json", action="store_true")
    args = parser.parse_args()

    with tempfile.TemporaryDirectory() as tmp:
        full_bodies = install(os.path.join(tmp, "full"), simple=False)
        simple_bodies = install(os.path.join(tmp, "simple"), simple=True)

    if not full_bodies:
        print("no skills installed; run from the repo root", file=sys.stderr)
        return 1

    sources = {
        os.path.basename(os.path.dirname(path)): open(path, encoding="utf-8").read()
        for path in sorted(glob.glob("cmd/magus/skills/*/SKILL.md"))
    }

    rows = []
    for name, full in sorted(full_bodies.items()):
        simple = simple_bodies.get(name, "")
        body = sources.get(name, "")
        comp = compose(simple)
        hard = comp["table"] + comp["code"]
        total = sum(comp.values()) or 1
        rows.append({
            "skill": name,
            "full_bytes": len(full),
            "simple_bytes": len(simple),
            "cut_pct": round(100 - 100 * len(simple) / len(full), 1) if full else 0.0,
            "full_only_branches": body.count("{{if .Full}}"),
            "simple_only_branches": body.count("{{if .Simple}}"),
            "two_wordings": body.count("{{else}}"),
            "simple_prose_pct": round(100 * comp["prose"] / total, 1),
            "simple_hard_pct": round(100 * hard / total, 1),
        })

    tot_full = sum(row["full_bytes"] for row in rows)
    tot_simple = sum(row["simple_bytes"] for row in rows)
    summary = {
        "skills": rows,
        "total_full_bytes": tot_full,
        "total_simple_bytes": tot_simple,
        "total_cut_pct": round(100 - 100 * tot_simple / tot_full, 1),
        "total_full_only_branches": sum(row["full_only_branches"] for row in rows),
        "total_simple_only_branches": sum(row["simple_only_branches"] for row in rows),
        "total_two_wordings": sum(row["two_wordings"] for row in rows),
    }

    if args.json:
        print(json.dumps(summary, indent=2))
        return 0

    print("{:<22} {:>7} {:>7} {:>6} {:>6} {:>6} {:>8}".format(
        "skill", "full", "simple", "cut", "ifFull", "else", "prose%"))
    for row in rows:
        print("{:<22} {:>7} {:>7} {:>5}% {:>6} {:>6} {:>7}%".format(
            row["skill"], row["full_bytes"], row["simple_bytes"],
            row["cut_pct"], row["full_only_branches"], row["two_wordings"],
            row["simple_prose_pct"]))
    print("{:<22} {:>7} {:>7} {:>5}% {:>6} {:>6}".format(
        "TOTAL", summary["total_full_bytes"], summary["total_simple_bytes"],
        summary["total_cut_pct"], summary["total_full_only_branches"],
        summary["total_two_wordings"]))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: evals/tools/measure_skills.mjs
// Report compression stats for the embedded agent skills.
//
// Installs BOTH permutations with the real renderer into temporary directories
// and measures the flat files that actually ship.
//
// Going through `magus agent install` rather than re-implementing applyVariant
// here is deliberate: this script cannot drift from the Go renderer, and it
// keeps working unchanged when the body syntax changes.
//
// Usage: node tools/measure_skills.mjs [--json]
//
// MAGUS_BIN overrides which magus renders (e.g. MAGUS_BIN=./magus for a HEAD
// build); the default is the magus on PATH.

import { execFileSync } from "node:child_process";
import { globSync, mkdtempSync, readFileSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { basename, dirname, join } from "node:path";
import process from "node:process";

// Paths are resolved from this file, not the caller's cwd, so the script works
// no matter where it is invoked from.
const ROOT = new URL("../../", import.meta.url).pathname;
const MAGUS = (process.env.MAGUS_BIN ?? "magus").split(/\s+/);

// Render one permutation into dest with the real renderer.
function install(dest, simple) {
  const args = [...MAGUS.slice(1), "agent", "install", "--dir", dest, "--force", ".claude/skills"];
  if (simple) {
    args.push("--simple");
  }
  // The renderer narrates every file it writes; keep that out of the report
  // and surface it only when the install actually fails.
  try {
    execFileSync(MAGUS[0], args, { cwd: ROOT, stdio: ["ignore", "ignore", "pipe"] });
  } catch (error) {
    throw new Error(`install failed: ${[MAGUS[0], ...args].join(" ")}\n${error.stderr ?? error.message}`);
  }
  const bodies = {};
  for (const path of globSync(join(dest, ".claude", "skills", "*", "SKILL.md")).sort()) {
    bodies[basename(dirname(path))] = readFileSync(path, "utf8");
  }
  return bodies;
}

// Byte counts by line category, for the prose vs hard-content split.
function compose(text) {
  const counts = { prose: 0, table: 0, code: 0, heading: 0, blank: 0 };
  let inFence = false;
  for (const line of text.split("\n")) {
    const n = line.length + 1;
    const stripped = line.trim();
    if (stripped.startsWith("```")) {
      inFence = !inFence;
      counts.code += n;
    } else if (inFence) {
      counts.code += n;
    } else if (stripped.startsWith("|")) {
      counts.table += n;
    } else if (stripped.startsWith("#")) {
      counts.heading += n;
    } else if (stripped === "") {
      counts.blank += n;
    } else {
      counts.prose += n;
    }
  }
  return counts;
}

function count(haystack, needle) {
  return haystack.split(needle).length - 1;
}

function round1(number) {
  return Math.round(number * 10) / 10;
}

function main() {
  const asJson = process.argv.includes("--json");

  const tmp = mkdtempSync(join(tmpdir(), "measure-skills-"));
  let fullBodies;
  let simpleBodies;
  try {
    fullBodies = install(join(tmp, "full"), false);
    simpleBodies = install(join(tmp, "simple"), true);
  } finally {
    rmSync(tmp, { recursive: true, force: true });
  }

  if (Object.keys(fullBodies).length === 0) {
    process.stderr.write("no skills installed; is magus on PATH?\n");
    return 1;
  }

  const sources = {};
  for (const path of globSync(join(ROOT, "cmd", "magus", "skills", "*", "SKILL.md")).sort()) {
    sources[basename(dirname(path))] = readFileSync(path, "utf8");
  }

  const rows = [];
  for (const name of Object.keys(fullBodies).sort()) {
    const full = fullBodies[name];
    const simple = simpleBodies[name] ?? "";
    const body = sources[name] ?? "";
    const comp = compose(simple);
    const hard = comp.table + comp.code;
    const total = Object.values(comp).reduce((sum, v) => sum + v, 0) || 1;
    rows.push({
      skill: name,
      full_bytes: full.length,
      simple_bytes: simple.length,
      cut_pct: full.length > 0 ? round1(100 - (100 * simple.length) / full.length) : 0,
      full_only_branches: count(body, "{{if .Full}}"),
      simple_only_branches: count(body, "{{if .Simple}}"),
      two_wordings: count(body, "{{else}}"),
      simple_prose_pct: round1((100 * comp.prose) / total),
      simple_hard_pct: round1((100 * hard) / total),
    });
  }

  const totFull = rows.reduce((sum, row) => sum + row.full_bytes, 0);
  const totSimple = rows.reduce((sum, row) => sum + row.simple_bytes, 0);
  const summary = {
    skills: rows,
    total_full_bytes: totFull,
    total_simple_bytes: totSimple,
    total_cut_pct: round1(100 - (100 * totSimple) / totFull),
    total_full_only_branches: rows.reduce((sum, row) => sum + row.full_only_branches, 0),
    total_simple_only_branches: rows.reduce((sum, row) => sum + row.simple_only_branches, 0),
    total_two_wordings: rows.reduce((sum, row) => sum + row.two_wordings, 0),
  };

  if (asJson) {
    console.log(JSON.stringify(summary, null, 2));
    return 0;
  }

  const cell = (value, width) => String(value).padStart(width);
  console.log(
    `${"skill".padEnd(22)} ${cell("full", 7)} ${cell("simple", 7)} ${cell("cut", 6)} ` +
      `${cell("ifFull", 6)} ${cell("else", 6)} ${cell("prose%", 8)}`,
  );
  for (const row of rows) {
    console.log(
      `${row.skill.padEnd(22)} ${cell(row.full_bytes, 7)} ${cell(row.simple_bytes, 7)} ` +
        `${cell(row.cut_pct.toFixed(1), 5)}% ${cell(row.full_only_branches, 6)} ` +
        `${cell(row.two_wordings, 6)} ${cell(row.simple_prose_pct.toFixed(1), 7)}%`,
    );
  }
  console.log(
    `${"TOTAL".padEnd(22)} ${cell(summary.total_full_bytes, 7)} ${cell(summary.total_simple_bytes, 7)} ` +
      `${cell(summary.total_cut_pct.toFixed(1), 5)}% ${cell(summary.total_full_only_branches, 6)} ` +
      `${cell(summary.total_two_wordings, 6)}`,
  );
  return 0;
}

process.exit(main());
